### ISSUE TITLE
Why MongoCollection::ensureDocumentHasMongoId() checks if $property is public?

### ISSUE THREAD

**Issue by mynthon** [alcaeus/mongo-php-adapter#252]
As method's name suggest this will set id for documents without it, but it checks every property if it is public throwing exception otherwise. Why it have to be checked?

**Comment by alcaeus** [comment]
The check is done to replicate behaviour from the legacy driver itself, which throws an exception when encountering a key that starts with a null-byte: https://github.com/mongodb/mongo-php-driver-legacy/blob/7dcbf423f0fe4f308d06bb49bff4c2f6e9e65657/bson.c#L593.

When casting an object to an array, which is essentially what happens when passing an object to `insert()`, private and protected keys are prefixed with null-bytes, which can be seen in the following 3v4l snippet: https://3v4l.org/euJBE

To properly replicate this behaviour of the legacy driver, the method that checks whether an object has an identifier also checks for any non-public properties, as they would have triggered this behaviour in the legacy driver. This is also checked in the tests with `MongoCollectionTest::testInsertObjectWithPrivateProperties`, seen in https://github.com/alcaeus/mongo-php-adapter/blob/b4c2f0d30dbddd624d3bc2a27175e6eb602241d4/tests/Alcaeus/MongoDbAdapter/Mongo/MongoCollectionTest.php#L89..L96.

Let me know if you have any further questions.

**Comment by mynthon** [comment]
Thank you for answer, but I still don't get why it must be checked in mthod which checks for id. I have object with private properties and custom __toArray() and I get exception everytime i am trying to save it, because of id check while (array)$myObj returns properly formatted data without null-bytes/zero-length keys.

When I comment out 

```
MongoCollection::ensureDocumentHasMongoId(){
...
    if (! $property->isPublic()) {
        throw new \MongoException('zero-length keys are not allowed, did you use $ with double quotes?');
    }
...
}
```

everything work ok. Not sure how to avoid this problem.

**Comment by alcaeus** [comment]
If you can provide a small sample class that reproduces the problem I'm happy to take a look.

**Comment by mynthon** [comment]
It is not real code but it gives in idea:

```
MyMongoDocument extends ArrayObject {
   protected $_log;
   protected $_collection;

    public static __contruct(MongoCollection $collection, $log) {
        $this->_collection = $collection;
        $this->_log = $log;
    }

    // (...)

    public static function save() {
        $this->_collection->insert($this);
    }
}
```

**Comment by alcaeus** [comment]
Fixed in 1.1.8. Thanks for bringing this to my attention!
